3.8-SANPSHOT Aggregation count test stable cluster large load,  AssertionException

while running an Aggregation count test, i got 
 hz.map.aggregate.Count threadId=0 global.AssertionException: 1026002168464627!=100000

I was using a stable cluster of 2 Members and 400 clients

```
Dec 28, 2016 1:53:30 PM com.hazelcast.map.impl.query.QueryOperation
SEVERE: [10.212.40.109]:5701 [HZ] [3.8-SNAPSHOT] java.util.concurrent.TimeoutException: HDPartitionScanOperation failed to complete within 57584181128 NANOSECONDS. Invocation{op=com
.hazelcast.map.impl.query.HDPartitionScanOperation{serviceName='hz:impl:mapService', identityHash=1027133611, partitionId=7, replicaIndex=0, callId=251558, invocationTime=1482925918
474 (2016-12-28 13:51:58.474), waitTimeout=-1, callTimeout=60000, name=mapBak1HDIdx}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs
=1482925921714, firstInvocationTime='2016-12-28 13:52:01.714', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.40.109]:5701, pendingResponse={VOID
}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
com.hazelcast.core.HazelcastException: java.util.concurrent.TimeoutException: HDPartitionScanOperation failed to complete within 57584181128 NANOSECONDS. Invocation{op=com.hazelcast
.map.impl.query.HDPartitionScanOperation{serviceName='hz:impl:mapService', identityHash=1027133611, partitionId=7, replicaIndex=0, callId=251558, invocationTime=1482925918474 (2016-
12-28 13:51:58.474), waitTimeout=-1, callTimeout=60000, name=mapBak1HDIdx}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=148292592
1714, firstInvocationTime='2016-12-28 13:52:01.714', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.40.109]:5701, pendingResponse={VOID}, backups
AcksExpected=0, backupsAcksReceived=0, connection=null}
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:73)
	at com.hazelcast.util.ExceptionUtil.peel(ExceptionUtil.java:52)
	at com.hazelcast.util.ExceptionUtil.rethrow(ExceptionUtil.java:83)
	at com.hazelcast.util.FutureUtil$1.handleException(FutureUtil.java:50)
	at com.hazelcast.util.FutureUtil.returnWithDeadline(FutureUtil.java:272)
	at com.hazelcast.util.FutureUtil.returnWithDeadline(FutureUtil.java:237)
	at com.hazelcast.map.impl.query.ParallelPartitionScanExecutor.waitForResult(ParallelPartitionScanExecutor.java:88)
	at com.hazelcast.map.impl.query.ParallelPartitionScanExecutor.runUsingPartitionScanWithoutPaging(ParallelPartitionScanExecutor.java:74)
	at com.hazelcast.map.impl.query.ParallelPartitionScanExecutor.execute(ParallelPartitionScanExecutor.java:56)
	at com.hazelcast.map.impl.query.ParallelPartitionScanExecutor.execute(ParallelPartitionScanExecutor.java:41)
	at com.hazelcast.map.impl.query.QueryRunner.runUsingPartitionScanSafely(QueryRunner.java:153)
	at com.hazelcast.map.impl.query.QueryRunner.run(QueryRunner.java:90)
	at com.hazelcast.map.impl.query.QueryOperation.run(QueryOperation.java:48)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:186)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:373)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:400)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:532)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:517)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:488)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:202)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59)
	at com.hazelcast.client.impl.protocol.task.map.AbstractMapQueryMessageTask.createInvocations(AbstractMapQueryMessageTask.java:122)
	at com.hazelcast.client.impl.protocol.task.map.AbstractMapQueryMessageTask.invokeOnMembers(AbstractMapQueryMessageTask.java:101)
	at com.hazelcast.client.impl.protocol.task.map.AbstractMapQueryMessageTask.call(AbstractMapQueryMessageTask.java:90)
	at com.hazelcast.client.impl.protocol.task.AbstractCallableMessageTask.processMessage(AbstractCallableMessageTask.java:35)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:123)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:103)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.Hazelcast
	ManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
	Caused by: java.util.concurrent.TimeoutException: HDPartitionScanOperation failed to complete within 57584181128 NANOSECONDS. Invocation{op=com.hazelcast.map.impl.query.HDPartitionS
canOperation{serviceName='hz:impl:mapService', identityHash=1027133611, partitionId=7, replicaIndex=0, callId=251558, invocationTime=1482925918474 (2016-12-28 13:51:58.474), waitTim
eout=-1, callTimeout=60000, name=mapBak1HDIdx}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1482925921714, firstInvocationTime='2
016-12-28 13:52:01.714', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.40.109]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksR
eceived=0, connection=null}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newTimeoutException(InvocationFuture.java:64)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:194)
	at com.hazelcast.util.FutureUtil.executeWithDeadline(FutureUtil.java:343)
	at com.hazelcast.util.FutureUtil.returnWithDeadline(FutureUtil.java:267)
	... 27 more
Add Comment Click to expand inline 54 lines
```

the above was from the Member which reported the assertion exception

but other Member showed slightly different exceptions

```
Dec 28, 2016 1:53:56 PM com.hazelcast.map.impl.query.MapQueryEngineImpl
WARNING: [10.212.40.108]:5701 [HZ] [3.8-SNAPSHOT] Could not get results
java.util.concurrent.ExecutionException: QueryOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2016-12-28 13:53:56.469. Start time: 2016-12-
28 13:51:56.226. Total elapsed time: 120243 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2016-12-28 13:53:34.784. Invocation{op=com.hazelcast.map.impl.
query.QueryOperation{serviceName='hz:impl:mapService', identityHash=1926086852, partitionId=-1, replicaIndex=0, callId=-222347, invocationTime=1482925916226 (2016-12-28 13:51:56.226
), waitTimeout=-1, callTimeout=60000, name=mapBak1}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1482925916226, firstInvocationTi
me='2016-12-28 13:51:56.226', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.40.109]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backups
AcksReceived=0, connection=Connection[id=2, /10.212.40.108:5701->/10.212.40.109:47411, endpoint=[10.212.40.109]:5701, alive=true, type=MEMBER]}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.newOperationTimeoutException(InvocationFuture.java:151)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolve(InvocationFuture.java:101)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationFuture.resolveAndThrowIfException(InvocationFuture.java:75)
	at com.hazelcast.spi.impl.AbstractInvocationFuture.get(AbstractInvocationFuture.java:155)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.addResultsOfPredicate(MapQueryEngineImpl.java:170)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.dispatchQueryOnQueryThreads(MapQueryEngineImpl.java:146)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.doRunQueryOnQueryThreads(MapQueryEngineImpl.java:139)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.runQueryOnAllPartitions(MapQueryEngineImpl.java:113)
	at com.hazelcast.map.impl.query.MapQueryEngineImpl.execute(MapQueryEngineImpl.java:87)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.aggregate(MapProxyImpl.java:783)
	at hz.map.aggregate.Count.timeStep(Count.java:15)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:52)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:39)
	at remote.bench.BenchThread.call(BenchThread.java:38)
	at remote.bench.BenchThread.call(BenchThread.java:12)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.hazelcast.core.OperationTimeoutException: QueryOperation invocation failed to complete due to operation-heartbeat-timeout. Current time: 2016-12-28 13:53:56.469. Star
t time: 2016-12-28 13:51:56.226. Total elapsed time: 120243 ms. Last operation heartbeat: never. Last operation heartbeat from member: 2016-12-28 13:53:34.784. Invocation{op=com.haz
elcast.map.impl.query.QueryOperation{serviceName='hz:impl:mapService', identityHash=1926086852, partitionId=-1, replicaIndex=0, callId=-222347, invocationTime=1482925916226 (2016-12
-28 13:51:56.226), waitTimeout=-1, callTimeout=60000, name=mapBak1}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1482925916226, f
irstInvocationTime='2016-12-28 13:51:56.226', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 02:00:00.000', target=[10.212.40.109]:5701, pendingResponse={VOID}, backupsAcksExp
ected=0, backupsAcksReceived=0, connection=Connection[id=2, /10.212.40.108:5701->/10.212.40.109:47411, endpoint=[10.212.40.109]:5701, alive=true, type=MEMBER]}
	... 21 more
Add Comment Click to expand inline 36 lines
```

Tom Bujok 
 how much data do you have in the map? it's possible that the default operation timeout is not enough

Danny Conlon  
100000 Person objects which are small (edited)

Tom Bujok 
Does each out of the 400 clients execute the query all the time?
 
If so, it would be good to extend the timeout

https://github.com/tombujok/hazelcast/blob/41285556e5a1dee6e955487023cb7e16ed555a5f/hazelcast/src/main/java/com/hazelcast/spi/properties/GroupProperty.java#L184-L184 (edited)

Where's the place where the aggregation exception was thrown?


Danny Conlon
yes each of the clients and the members execute the query from multi thread at the same time
and each time the result is checked,
the time out are only in the logs of the member,  they are not thrown back to the caller
its ok if there are time out, as we are seeing what happens under heave load
but getting a wrong answer from the query, is the issue

Tom Bujok
aha, so the assertion error was due to a wrong aggregation result

Danny Conlon 
yes

Tom Bujok 
no bouncing members or anything like this - stable cluster, heavy load and then the assertion error occured?

Danny Conlon 
yes

the same test under cluster shutdown conditions is https://hazelcast-l337.ci.cloudbees.com/view/shutdown/job/shutdown-aggregate-count/  and it runs ok,  but only 4 members and 4 clients, and only clients making operations

Tom Bujok 
ok, is there a full log of the assertion failure? just to have a look what happened before/after? 
I mean on the members

Danny Conlon
the members and clients just run https://github.com/Danny-Hazelcast/hzCmd-bench/blob/master/src/main/java/hz/map/aggregate/Count.java
 GitHub
Danny-Hazelcast/hzCmd-bench
hzCmd-bench - bench mark implementation 
 
 
the 2 snippets posted above, were from 2 diff members.  and the logs are full of these repeated exceptions

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java
@@ -222,7 +222,7 @@ class MapServiceContextImpl implements MapServiceContext {
             accumulationExecutor = new CallerRunsAccumulationExecutor(ss);
         }
 
-        return new AggregationResultProcessor(accumulationExecutor);
+        return new AggregationResultProcessor(accumulationExecutor, nodeEngine.getSerializationService());
     }
 
     private PartitionContainer[] createPartitionContainers() {

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/query/AggregationResultProcessor.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/query/AggregationResultProcessor.java
@@ -16,7 +16,9 @@
 
 package com.hazelcast.map.impl.query;
 
+import com.hazelcast.aggregation.Aggregator;
 import com.hazelcast.query.impl.QueryableEntry;
+import com.hazelcast.spi.serialization.SerializationService;
 
 import java.util.Collection;
 
@@ -26,9 +28,11 @@ import java.util.Collection;
 public class AggregationResultProcessor implements ResultProcessor<AggregationResult> {
 
     private final AccumulationExecutor accumulationExecutor;
+    private final SerializationService serializationService;
 
-    public AggregationResultProcessor(AccumulationExecutor accumulationExecutor) {
+    public AggregationResultProcessor(AccumulationExecutor accumulationExecutor, SerializationService serializationService) {
         this.accumulationExecutor = accumulationExecutor;
+        this.serializationService = serializationService;
     }
 
     @Override
@@ -39,6 +43,7 @@ public class AggregationResultProcessor implements ResultProcessor<AggregationRe
 
     @Override
     public AggregationResult populateResult(Query query, long resultLimit) {
-        return new AggregationResult(query.getAggregator());
+        Aggregator resultAggregator = serializationService.toObject(serializationService.toData(query.getAggregator()));
+        return new AggregationResult(resultAggregator);
     }
 }

